src/graph/sampler.cc

```diff
@@ -111,6 +113,257 @@ class ArrayHeap {
   std::vector<ValueType> heap_;
 };
 
+/*
+ * AliasSampler is used to sample elements from a given discrete categorical distribution.
+ * Algorithm: Alias Method(https://en.wikipedia.org/wiki/Alias_method)
+ * Sampler building complexity: O(n)
+ * Sample w/ replacement complexity: O(1)
+ * Sample w/o replacement complexity: O(log n)
+ */
+template <
+    typename Idx,
+    typename DType,
+    bool replace>
+class AliasSampler {
+ private:
+  Idx N;
+  DType accum, taken, eps;
+  std::vector<Idx> K;             // alias table
+  std::vector<DType> U;           // probability table
+  std::vector<DType> _prob;       // categorical distribution
+  std::vector<bool> used;         // indicate availability, activated when replace=false;
+  std::vector<Idx> id_mapping;    // index mapping, activated when replace=false;
+
+  inline Idx map(Idx x) const {
+    if (replace)
+      return x;
+    else
+      return id_mapping[x];
+  }
+
+  void rebuild(const std::vector<DType>& prob) {
+    N = 0;
+    accum = 0.;
+    taken = 0.;
+    if (!replace)
+      id_mapping.clear();
+    for (Idx i = 0; i < prob.size(); ++i)
+      if (!used[i]) {
+        N++;
+        accum += prob[i];
+        if (!replace)
+          id_mapping.push_back(i);
+      }
+    if (N == 0) LOG(FATAL) << "Cannot take more sample than population when 'replace=false'";
+    K.resize(N);
+    U.resize(N);
+    DType avg = accum / static_cast<DType>(N);
+    std::fill(U.begin(), U.end(), avg);   // initialize U
+    std::queue<std::pair<Idx, DType> > under, over;
+    for (Idx i = 0; i < N; ++i) {
+      DType p = prob[map(i)];
+      if (p > avg)
+        over.push(std::make_pair(i, p));
+      else
+        under.push(std::make_pair(i, p));
+      K[i] = i;                           // initialize K
+    }
+    while (!under.empty() && !over.empty()) {
+      auto u_pair = under.front(), o_pair = over.front();
+      Idx i_u = u_pair.first, i_o = o_pair.first;
+      DType p_u = u_pair.second, p_o = o_pair.second;
+      K[i_u] = i_o;
+      U[i_u] = p_u;
+      if (p_o + p_u > 2 * avg + eps)
+        over.push(std::make_pair(i_o, p_o + p_u - avg));
+      else if (p_o + p_u < 2 * avg - eps)
+        under.push(std::make_pair(i_o, p_o + p_u - avg));
+      under.pop();
+      over.pop();
+    }
+  }
+
+ public:
+  void reinit_state(const std::vector<DType>& prob) {
+    used.resize(prob.size());
+    if (!replace)
+      _prob = prob;
+    std::fill(used.begin(), used.end(), false);
+    rebuild(prob);
+  }
+
+  explicit AliasSampler(const std::vector<DType>& prob, DType eps = 1e-12): eps(eps) {
+    reinit_state(prob);
+  }
+
+  ~AliasSampler() {}
+
+  inline Idx draw(RandomEngine *re) {
```

Comment:
I will prefer put the RandomEngine when initializing the Sampler. 
This will simplify the caller

Reply:
DONE